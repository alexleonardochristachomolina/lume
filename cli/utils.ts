import lume from "../mod.ts";
import { toFileUrl } from "../deps/path.ts";
import { isUrl } from "../core/utils/path.ts";
import { getConfigFile } from "../core/utils/lume_config.ts";
import { log } from "../core/utils/log.ts";
import Site from "../core/site.ts";

/** Create a site instance */
export async function createSite(config?: string): Promise<Site> {
  let url: string | undefined;
  const watchInclude = [];

  if (config && isUrl(config)) {
    url = config;
  } else {
    const path = await getConfigFile(config);

    if (path) {
      watchInclude.push(path);
      url = toFileUrl(path).href;
    }
  }

  if (url) {
    log.info(`Loading config file <gray>${url}</gray>`);
    const mod = await import(url);
    const site = mod.default as Site | undefined;
    if (!(site instanceof Site)) {
      log.critical(
        `[Lume] Missing Site instance! Ensure your config file does export the Site instance as default.`,
      );
      throw new Error("Site instance is not found");
    }

    site._data.configFile = url;
    site.options.watcher.include.push(...watchInclude);
    return site;
  }

  return lume();
}

/** Create a site intance and build it */
export async function buildSite(config?: string): Promise<Site> {
  const site = await createSite(config);

  performance.mark("start");
  await site.build();
  performance.mark("end");

  log.info(`üçæ Site built into <gray>${site.options.dest}</gray>`);
  const duration = performance.measure("duration", "start", "end").duration /
    1000;
  const total = site.pages.length + site.files.length;
  log.info(
    `  <gray>${total} files generated in ${duration.toFixed(2)} seconds</gray>`,
  );

  return site;
}
