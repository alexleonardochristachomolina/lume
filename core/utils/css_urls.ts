type UrlWalker = (url: string, type: string) => string | Promise<string>;
type Position = [string, string, number, number];
type Index = [string, number];

export async function walkUrls(
  code: string,
  walker: UrlWalker,
): Promise<string> {
  let result = "";
  let index = 0;
  let position = next(code, index);

  while (position) {
    const [type, url, start, end] = position;
    const newUrl = await walker(url, type);
    result += `${code.slice(index, start)}"${newUrl}"`;
    index = end;
    position = next(code, index);
  }

  return result + code.slice(index);
}

function next(code: string, index = 0): Position | undefined {
  const nextUrl = code.indexOf("url(", index);
  const nextComment = code.indexOf("/*", index);
  const nextImport = code.indexOf("@import ", index);

  let found: Index | undefined;

  if (nextUrl !== -1) {
    found = ["url", nextUrl];
  }
  if (nextComment !== -1 && (!found || nextComment < found[1])) {
    found = ["comment", nextComment];
  }
  if (nextImport !== -1 && (!found || nextImport < found[1])) {
    found = ["import", nextImport];
  }

  if (!found) {
    return;
  }

  // If it's a comment, find the end of it
  if (found[0] === "comment") {
    const end = code.indexOf("*/", found[1]);
    if (end === -1) {
      return;
    }
    return next(code, end + 2);
  }

  if (found[0] === "import") {
    const double = code.indexOf('"', found[1]);
    const single = code.indexOf("'", found[1]);
    const urlFn = code.indexOf("url(", found[1]);

    let foundImport: Index | undefined;

    if (double !== -1) {
      foundImport = ["double", double];
    }
    if (single !== -1 && (!foundImport || single < foundImport[1])) {
      foundImport = ["single", single];
    }

    // It's has a url(), so delegate to the url() logic
    if (urlFn !== -1 && (!foundImport || urlFn < foundImport[1])) {
      return next(code, urlFn);
    }

    if (!foundImport) {
      return;
    }

    const start = foundImport[1];
    const end =
      code.indexOf(foundImport[0] === "double" ? '"' : "'", start + 1) + 1;
    return ["url", code.slice(start + 1, end - 1), start, end];
  }

  if (found[0] === "url") {
    const start = code.indexOf("(", found[1]) + 1;
    const end = code.indexOf(")", start);
    let url = code.slice(start, end).trim();
    // Remove quotes
    if (url.startsWith('"') || url.startsWith("'")) {
      url = url.slice(1, -1);
    }
    return ["url", url, start, end];
  }
}
