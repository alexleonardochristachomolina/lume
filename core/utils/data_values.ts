import type { Data } from "../file.ts";

/**
 * Get the value of a page data
 * For example, if the value is "=title", it returns the value of the page data "title"
 * If the value is "$.title", it will return the value of the element with the selector ".title"
 */
export function getDataValue(data: Partial<Data>, value?: unknown) {
  // Get the value from the page data
  if (typeof value === "string") {
    return searchValue(data, value);
  }

  if (typeof value === "function") {
    return value(data);
  }

  return value;
}

function searchValue(data: Partial<Data>, value: string): unknown {
  if (!value) {
    return;
  }

  if (value.startsWith("=")) {
    let key = value.slice(1);
    [key, value] = parseFallback(key);

    if (!key.includes(".")) {
      return data[key] ?? searchValue(data, value);
    }

    const keys = key.split(".");
    let val = data;
    for (const key of keys) {
      val = val?.[key];
    }
    return val ?? searchValue(data, value);
  }

  if (value.startsWith("$")) {
    let selector = value.slice(1);
    [selector, value] = parseFallback(selector);

    return queryCss(selector, data.page?.document) ?? searchValue(data, value);
  }

  return value;
}

function parseFallback(key: string): [string, string] {
  const fallback = key.indexOf("||");

  if (fallback !== -1) {
    return [
      key.slice(0, fallback).trim(),
      key.slice(fallback + 2).trim(),
    ];
  }

  return [
    key,
    "",
  ];
}

// https://regexr.com/7qnot
const checkForAttrPattern = /^(.+)\s+(?:attr\(([\w\-]+)\))$/;

function queryCss(query: string, document?: Document) {
  const checkResult = query.match(checkForAttrPattern);

  const hasAttr = checkResult?.[0];
  if (hasAttr) {
    const [_, query, name] = checkResult;
    return document?.querySelector(query)?.getAttribute(name);
  }

  return document?.querySelector(query)?.innerHTML;
}
