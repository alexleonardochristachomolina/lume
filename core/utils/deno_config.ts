import { parse } from "../../deps/jsonc.ts";
import { isUrl } from "../utils/path.ts";

/** Import map file */
export interface ImportMap {
  imports: Record<string, string>;
  scopes?: Record<string, Record<string, string>>;
}

/** Basic options for deno.json file */
export interface DenoConfig extends Partial<ImportMap> {
  importMap?: string;
  tasks?: Record<string, string>;
  compilerOptions?: {
    jsx?: "jsx" | "react-jsx" | "precompile";
    jsxImportSource?: string;
    types?: string[];
  };
  [key: string]: unknown;
}

export interface DenoConfigResult {
  file: string;
  config: DenoConfig;
  importMap?: ImportMap;
}

/** Detect and returns the Deno configuration */
export async function readDenoConfig(): Promise<DenoConfigResult | undefined> {
  for (const file of ["deno.json", "deno.jsonc"]) {
    try {
      const content = await Deno.readTextFile(file);
      const config = parse(content) as DenoConfig;
      let importMap: ImportMap | undefined;

      if (config.importMap) {
        importMap = isUrl(config.importMap)
          ? await (await fetch(config.importMap)).json()
          : await JSON.parse(await Deno.readTextFile(config.importMap));
      } else if (config.imports) {
        importMap = {
          imports: config.imports,
          scopes: config.scopes,
        };
      }
      return { file, config, importMap };
    } catch (err) {
      if (err instanceof Deno.errors.NotFound) {
        continue;
      }

      throw err;
    }
  }
}
