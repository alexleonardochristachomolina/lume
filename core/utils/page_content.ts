import type { Page } from "../file.ts";

export function insertContent(
  page: Page,
  content: string,
  placeholder?: string,
) {
  const pageContent = page.content;

  if (pageContent && typeof pageContent === "string") {
    if (placeholder && pageContent.includes(placeholder)) {
      page.content = pageContent.replace(
        placeholder,
        content,
      );
    } else {
      page.content += `\n${content}`;
    }
  } else {
    page.content = content;
  }
}
