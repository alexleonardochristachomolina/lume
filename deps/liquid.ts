export * from "npm:liquidjs@10.19.0";

export type { LiquidOptions, Template } from "npm:liquidjs@10.19.0";
export type {
  TagClass,
  TagImplOptions,
} from "npm:liquidjs@10.19.0/dist/template";
