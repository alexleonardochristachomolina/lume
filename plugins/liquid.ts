import {
  evalToken,
  Liquid,
  Tag,
  Tokenizer,
  TokenKind,
  toPromise,
  Value,
  ValueToken,
} from "../deps/liquid.ts";
import { posix } from "../deps/path.ts";
import loader from "../core/loaders/text.ts";
import { merge } from "../core/utils/object.ts";

import type { Data } from "../core/file.ts";
import type Site from "../core/site.ts";
import type { Engine, Helper, HelperOptions } from "../core/renderer.ts";
import type {
  Context,
  Emitter,
  LiquidOptions,
  TagClass,
  TagToken,
  Template,
  TopLevelToken,
} from "../deps/liquid.ts";

export interface Options {
  /** The list of extensions this plugin applies to */
  extensions?: string[];

  /** Optional sub-extension for page files */
  pageSubExtension?: string;

  /**
   * Custom includes path
   * @default `site.options.includes`
   */
  includes?: string;

  /**
   * Options passed to Liquidjs library
   * @see https://liquidjs.com/tutorials/options.html
   */
  options?: LiquidOptions;
}

// Default options
export const defaults: Options = {
  extensions: [".liquid"],
};

/** Template engine to render Liquid files */
export class LiquidEngine implements Engine {
  liquid: Liquid;
  cache = new Map<string, Template[]>();
  basePath: string;
  includes: string;

  constructor(liquid: Liquid, basePath: string, includes: string) {
    this.liquid = liquid;
    this.basePath = basePath;
    this.includes = includes;
  }

  deleteCache(file: string): void {
    this.cache.delete(file);
  }

  async render(
    content: string,
    data?: Record<string, unknown>,
    filename?: string,
  ) {
    if (!filename) {
      return this.liquid.parseAndRender(content, data);
    }

    const template = this.getTemplate(content, filename);
    return await this.liquid.render(template, data);
  }

  renderComponent(
    content: string,
    data?: Record<string, unknown>,
    filename?: string,
  ) {
    if (!filename) {
      return this.liquid.parseAndRenderSync(content, data);
    }

    const template = this.getTemplate(content, filename);
    return this.liquid.renderSync(template, data);
  }

  getTemplate(content: string, filename: string): Template[] {
    if (!this.cache.has(filename)) {
      this.cache.set(
        filename,
        this.liquid.parse(content, posix.join(this.basePath, filename)),
      );
    }
    return this.cache.get(filename)!;
  }

  addHelper(name: string, fn: Helper, options: HelperOptions) {
    switch (options.type) {
      case "filter":
        // deno-lint-ignore no-explicit-any
        this.liquid.registerFilter(name, function (this: any, ...args) {
          return fn.apply({ data: this.context.environments }, args);
        });
        break;

      case "tag":
        // Tag with body not supported yet
        if (!options.body) {
          this.liquid.registerTag(name, createCustomTag(fn));
        } else {
          this.liquid.registerTag(name, createCustomTagWithBody(fn));
        }
        break;
    }
  }
}

/**
 * A plugin to render Liquid files
 * @see https://lume.land/plugins/liquid/
 * @deprecated Use Vento or Nunjucks instead (see https://github.com/lumeland/lume/issues/600)
 */
export function liquid(userOptions?: Options) {
  return function (site: Site) {
    const options = merge(
      { ...defaults, includes: site.options.includes },
      userOptions,
    );

    const liquidOptions: LiquidOptions = {
      root: site.src(options.includes),
      ...options.options,
    };

    const engine = new LiquidEngine(
      new Liquid(liquidOptions),
      site.src(),
      options.includes,
    );

    // Ignore includes folder
    if (options.includes) {
      site.ignore(options.includes);
    }

    // Load the pages and register the engine
    site.loadPages(options.extensions, {
      loader,
      engine,
      pageSubExtension: options.pageSubExtension,
    });

    // Register the liquid filter
    site.filter("liquid", filter, true);

    function filter(
      string: string,
      data?: Record<string, unknown>,
    ): Promise<string> {
      return engine.render(string, { ...site.scopedData.get("/"), ...data });
    }
  };
}

/**
 * Create a custom tag
 * https://liquidjs.com/tutorials/register-filters-tags.html#Register-Tags
 */
function createCustomTag(fn: Helper): TagClass {
  return class extends Tag {
    #value: Value;

    constructor(
      token: TagToken,
      remainTokens: TopLevelToken[],
      liquid: Liquid,
    ) {
      super(token, remainTokens, liquid);
      this.#value = new Value(token.args, liquid);
    }

    async render(ctx: Context, emitter: Emitter) {
      const str = await toPromise(this.#value.value(ctx, false));
      emitter.write(await fn.call({ data: ctx.environments as Data }, str));
    }
  };
}

/**
 * Create a custom tag with body
 * https://liquidjs.com/tutorials/register-filters-tags.html#Register-Tags
 */
function createCustomTagWithBody(fn: Helper): TagClass {
  return class extends Tag {
    args: ValueToken[] = [];
    templates: Template[] = [];

    constructor(
      token: TagToken,
      remainTokens: TopLevelToken[],
      liquid: Liquid,
    ) {
      super(token, remainTokens, liquid);
      const tokenizer = new Tokenizer(
        token.args,
        this.liquid.options.operators,
      );
      const name = token.name;

      while (!tokenizer.end()) {
        const value = tokenizer.readValue();
        if (value) this.args.push(value);
        tokenizer.readTo(",");
      }

      while (remainTokens.length) {
        const token = remainTokens.shift()!;
        if (
          token.kind === TokenKind.Tag &&
          (token as TagToken).name === `end${name}`
        ) {
          return;
        }
        this.templates.push(liquid.parser.parseToken(token, remainTokens));
      }
      throw new Error(`tag ${token.getText()} not closed`);
    }

    async render(ctx: Context, emitter: Emitter) {
      const r = this.liquid.renderer;
      const str = await toPromise(r.renderTemplates(this.templates, ctx));
      const args = [str];

      for (const arg of this.args) {
        args.push(await toPromise(evalToken(arg, ctx)));
      }

      emitter.write(await fn.apply({ data: ctx.environments as Data }, args));
    }
  };
}

export default liquid;

/** Extends Helpers interface */
declare global {
  namespace Lume {
    export interface Helpers {
      /** @see https://lume.land/plugins/liquid/ */
      liquid: (
        string: string,
        data?: Record<string, unknown>,
      ) => Promise<string>;
    }
  }
}
