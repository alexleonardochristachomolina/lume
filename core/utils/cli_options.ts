import { parseArgs } from "../../deps/cli.ts";
import type { DeepPartial } from "./object.ts";
import type { SiteOptions } from "../site.ts";

export function getOptionsFromCli(
  options: DeepPartial<SiteOptions>,
): DeepPartial<SiteOptions> {
  const cli = parseArgs(Deno.args, {
    string: ["src", "dest", "location", "port"],
    boolean: ["serve", "open"],
    alias: { dev: "d", serve: "s", port: "p", open: "o" },
    ["--"]: true,
  });

  if (cli.src) {
    options.src = cli.src;
  }

  if (cli.dest) {
    options.dest = cli.dest;
  }

  const serveMode = cli.serve || cli._[0] === "cms";

  if (cli.port) {
    (options.server ||= {}).port = parseInt(cli.port);
  } else if (serveMode) {
    (options.server ||= {}).port = 3000;
  }

  let location: URL;

  if (cli.location) {
    location = new URL(cli.location);
  } else if (options.location && !serveMode) {
    location = options.location as URL;
  } else {
    location = new URL("http://localhost");
  }

  let port: number;

  if (cli.port) {
    port = parseInt(cli.port);
  } else if (location.port) {
    port = parseInt(location.port);
  } else if (options.server?.port) {
    port = options.server.port;
  } else {
    port = cli.serve ? 3000 : location.protocol === "https:" ? 443 : 80;
  }

  (options.server ||= {}).port = port;
  location.port = port.toString();
  options.location = location;

  if (cli.open) {
    (options.server ||= {}).open = cli.open;
  }

  return options;
}
