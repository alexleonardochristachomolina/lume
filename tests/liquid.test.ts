import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import liquid from "../plugins/liquid.ts";

Deno.test("build a site with liquid", async (t) => {
  const site = getSite({
    src: "liquid",
    location: new URL("https://example.com/blog"),
  });

  site.use(liquid());

  // Register an async filter
  site.filter(
    "returnAsync",
    (text) =>
      new Promise((resolve) =>
        setTimeout(() => resolve(`${text} (async)`), 10)
      ),
    true,
  );

  // Register custom tags
  site.helper(
    "upperCase",
    (text) => `<strong>${(text as string).toUpperCase()}</strong>`,
    { type: "tag" },
  );
  site.helper(
    "upperCaseBody",
    (text) => `<strong>${(text as string).toUpperCase()}</strong>`,
    { type: "tag", body: true },
  );
  site.helper(
    "upperCaseAsync",
    (text) =>
      Promise.resolve(`<strong>${(text as string).toUpperCase()}</strong>`),
    { type: "tag", async: true },
  );
  site.helper(
    "upperCaseBodyAsync",
    (text) =>
      Promise.resolve(`<strong>${(text as string).toUpperCase()}</strong>`),
    { type: "tag", body: true, async: true },
  );

  site.filter("fromPage", function (key) {
    return this?.data[key];
  });
  site.filter("fromPageAsync", function (key) {
    return Promise.resolve(this?.data[key]);
  }, true);
  site.helper(
    "fromPageTagAsync",
    function (key) {
      return Promise.resolve(`<strong>${this?.data[key]}</strong>`);
    },
    { type: "tag", body: true, async: true },
  );

  await build(site);
  await assertSiteSnapshot(t, site);
});
