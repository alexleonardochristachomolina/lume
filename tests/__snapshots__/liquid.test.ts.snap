export const snapshot = {};

snapshot[`build a site with liquid 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      loader: [AsyncFunction: toml],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.ts",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.js",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.json",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".markdown",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".vto",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".liquid",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
  ],
  src: [
    "/",
    "/_data.yml",
    "/_includes",
    "/_includes/basic.liquid",
    "/_includes/partial.liquid",
    "/data.liquid",
    "/empty.liquid",
    "/index.liquid",
    "/liquid-filter.page.js",
    "/with-helpers.liquid",
  ],
}
`;

snapshot[`build a site with liquid 2`] = `[]`;

snapshot[`build a site with liquid 3`] = `
[
  {
    content: '["red","green","blue"]
',
    data: {
      basename: "data",
      colors: "Array(3)",
      content: "{{ colors | json }}
",
      date: [],
      layout: null,
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/data.json",
    },
    src: {
      asset: false,
      ext: ".liquid",
      path: "/data",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Hello World</title>
  </head>
  <body>
    <main><h1>Hello World</h1>
</main>
  </body>
</html>
',
    data: {
      basename: "index",
      children: "<h1>Hello World</h1>
",
      content: "<h1>{{ title }}</h1>
",
      date: [],
      layout: "basic.liquid",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Hello World",
      url: "/",
    },
    src: {
      asset: false,
      ext: ".liquid",
      path: "/index",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>liquid filter example</title>
  </head>
  <body>
    <main><h1>LIQUID FILTER EXAMPLE</h1></main>
  </body>
</html>
',
    data: {
      basename: "liquid-filter",
      children: "<h1>LIQUID FILTER EXAMPLE</h1>",
      content: "default",
      date: [],
      layout: "basic.liquid",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "liquid filter example",
      url: "/liquid-filter/",
    },
    src: {
      asset: false,
      ext: ".page.js",
      path: "/liquid-filter",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>The title</title>
  </head>
  <body>
    <main><p><strong>THE TITLE</strong></p>
<p><strong>THE TITLE</strong></p>
<p><strong>THE TITLE</strong></p>
<p><strong>THE TITLE</strong></p>
<p><strong>THE TITLE</strong></p>
<p><strong>THE TITLE</strong></p>
<p><strong>THE TITLE</strong></p>
<p><strong>THE TITLE</strong></p>
<div>hello (async)</div>

<ul>
  <li>This is a partial</li>
  <li>async helper in a partial (async)</li>
</ul>
<ul>
  <li>This is a partial</li>
  <li>async helper in a partial (async)</li>
</ul>

<strong>The title</strong>
<strong>The title</strong>
<p><strong>The title</strong></p>
</main>
  </body>
</html>
',
    data: {
      basename: "with-helpers",
      children: "<p><strong>THE TITLE</strong></p>
<p><strong>THE TITLE</strong></p>
<p><strong>THE TITLE</strong></p>
<p><strong>THE TITLE</strong></p>
<p><strong>THE TITLE</strong></p>
<p><strong>THE TITLE</strong></p>
<p><strong>THE TITLE</strong></p>
<p><strong>THE TITLE</strong></p>
<div>hello (async)</div>

<ul>
  <li>This is a partial</li>
  <li>async helper in a partial (async)</li>
</ul>
<ul>
  <li>This is a partial</li>
  <li>async helper in a partial (async)</li>
</ul>

<strong>The title</strong>
<strong>The title</strong>
<p><strong>The title</strong></p>
",
      content: '<p>{% upperCase title %}</p>
<p>{% upperCase "The title" %}</p>
<p>{% upperCaseAsync title %}</p>
<p>{% upperCaseAsync "The title" %}</p>
<p>{% upperCaseBody %}The title{% endupperCaseBody %}</p>
<p>{% upperCaseBody %}{{ title }}{% endupperCaseBody %}</p>
<p>{% upperCaseBodyAsync %}The title{% endupperCaseBodyAsync %}</p>
<p>{% upperCaseBodyAsync %}{{ title }}{% endupperCaseBodyAsync %}</p>
<div>{{ "hello" | returnAsync }}</div>

{% include "partial.liquid" %}
{% include "./_includes/partial.liquid" %}

<strong>{{ "title" | fromPage }}</strong>
<strong>{{ "title" | fromPageAsync }}</strong>
<p>{% fromPageTagAsync %}title{% endfromPageTagAsync %}</p>
',
      date: [],
      layout: "basic.liquid",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "The title",
      url: "/with-helpers/",
    },
    src: {
      asset: false,
      ext: ".liquid",
      path: "/with-helpers",
      remote: undefined,
    },
  },
]
`;
